sway/tree/workspace.c

```diff
@@ -730,19 +691,39 @@ void workspace_add_gaps(struct sway_workspace *ws) {
 			focus = seat_get_focus_inactive_view(seat, &focus->node);
 		}
 		if (focus && focus->view && view_is_only_visible(focus->view)) {
+			ws->current_gaps.top = 0;
+			ws->current_gaps.right = 0;
+			ws->current_gaps.bottom = 0;
+			ws->current_gaps.left = 0;
 			return;
 		}
 	}
 
 	ws->current_gaps = ws->gaps_outer;
-	if (ws->layout == L_TABBED || ws->layout == L_STACKED) {
-		// We have to add inner gaps for this, because children of tabbed and
-		// stacked containers don't apply their own gaps - they assume the
-		// tabbed/stacked container is using gaps.
-		ws->current_gaps.top += ws->gaps_inner;
-		ws->current_gaps.right += ws->gaps_inner;
-		ws->current_gaps.bottom += ws->gaps_inner;
-		ws->current_gaps.left += ws->gaps_inner;
+	// Add inner gaps and make sure we don't turn out negative
+	ws->current_gaps.top = MAX(0, ws->current_gaps.top + ws->gaps_inner);
+	ws->current_gaps.right = MAX(0, ws->current_gaps.right + ws->gaps_inner);
+	ws->
+current_gaps.bottom = MAX(0, ws->current_gaps.bottom + ws->gaps_inner);
+	ws->current_gaps.left = MAX(0, ws->current_gaps.left + ws->gaps_inner);
+
+	// Now that we have the total gaps calculated we may need to clamp them in
+	// case they've made the available area too small
+	if (ws->width - ws->current_gaps.left - ws->current_gaps.right < MIN_SANE_W &&
+	    ws->current_gaps.left + ws->current_gaps.right > 0) {
+		int total_gap = MAX(0, ws->width - MIN_SANE_W);
+		double left_gap_frac = ((double)ws->current_gaps.left /
+			((double)ws->current_gaps.left + (double)ws->current_gaps.right));
+		ws->current_gaps.left = left_gap_frac * total_gap;
+		ws->current_gaps.right = total_gap - ws->current_gaps.left;
+	}
+	if (ws->height - ws->current_gaps.top - ws->current_gaps.bottom < MIN_SANE_H &&
+	    ws->current_gaps.top + ws->current_gaps.bottom > 0) {
```

Comment:
Like above, there is mixed tabs and spaces here. Please use tabs and indent this line one additional tab to make it distinguishable from the body of the block.  Can you also move the `&&` to this line so the previous one is at the 80 character limit?